src/runtime_src/core/tools/xbtracer/src/app/launcher.cpp

```diff
@@ -0,0 +1,534 @@
+  if (launcher::getInstance().m_debug) {
+    std::ostringstream oss;
+    oss << "[" << launcher::getInstance().m_name.data() << "] D: ";
+    log_format(oss, args...);
+    std::cout << oss.str() << std::endl;
+  }
+}
+
+/*
+ * thread safe wrapper of putenv()
+ */
+void putenv_t(std::string& new_entry)
+{
+  std::lock_guard<std::mutex> lock(env_mutex);
+  #ifdef _WIN32
+  _putenv(new_entry.data());
+  #else
+  // NOLINTNEXTLINE(concurrency-mt-unsafe) - protected by env_mutex
+  putenv(new_entry.data());
+  #endif
+}
+
+#ifdef _WIN32
+/*
+ * Check the architectural compatability(32bit vs 64bit) between the app and
+ * lib.
+ */
+static bool check_compatibility(HANDLE parent, HANDLE child)
+{
+  static launcher& app = launcher::getInstance();
+  BOOL is_parent_wow64 = FALSE, is_child_wow64 = FALSE;
+
+  IsWow64Process(parent, &is_parent_wow64);
+  IsWow64Process(child, &is_child_wow64);
+
+  if (is_parent_wow64 != is_child_wow64) {
+    log_e(app.m_name.data(), "is", (is_parent_wow64 ? w64 : w32),
+          "-bit but target application is ", (is_child_wow64 ? w64 : w32),
+          "-bit");
+    return false;
+  }
+
+  return true;
+}
+
+// Function to set an environment variable
+bool SetEnvVariable(const std::string& name, const std::string& value)
+{
+  LPCSTR lpName = name.c_str();
+  LPCSTR lpValue = value.c_str();
+
+  BOOL result = SetEnvironmentVariableA(lpName, lpValue);
+
+  if (result == 0) {
+    DWORD errorCode = GetLastError();
+    std::cerr << "Failed to set environment variable. Error code: " << errorCode
+              << std::endl;
+    return false;
+  }
+
+  return true;
+}
+
+/*
+ * Function to get the current precise system time as pair for string and
+ * formated string
+ */
+std::pair<std::string, std::string> GetCurrentTimeAsString()
+{
+  FILETIME ft;
+  GetSystemTimePreciseAsFileTime(&ft);
+
+  ULARGE_INTEGER uli;
+  uli.LowPart = ft.dwLowDateTime;
+  uli.HighPart = ft.dwHighDateTime;
+
+  // Convert the time to a string
+  std::ostringstream oss;
+  oss << uli.QuadPart;
+
+  // Time formatting and trace directory printing
+  FILETIME localFt;
+  FileTimeToLocalFileTime(&ft, &localFt);
+
+  SYSTEMTIME st;
+  FileTimeToSystemTime(&localFt, &st);
+
+  std::ostringstream ss;
+  ss << std::setfill('0') << std::setw(4) << st.wYear << "-" << std::setw(2)
+     << st.wMonth << "-" << std::setw(2) << st.wDay << "_" << std::setw(2)
+     << st.wHour << "-" << std::setw(2) << st.wMinute << "-" << std::setw(2)
+     << st.wSecond;
+
+  std::string formattedTime = ss.str();
+
+  return {oss.str(), formattedTime};
+}
+#else
+/* Warning: variable 'environ' is non-const and globally accessible
+ * It ensures compatibility with existing systems and libraries that rely on
+ * environ. Its usage is typically limited and encapsulated, reducing the risk
+ * of misuse.
+ */
+// NOLINTNEXTLINE (cppcoreguidelines-avoid-non-const-global-variables)
+extern char **environ;
+#endif
+
+/*
+ * Main Function
+ */
+int main(int argc, char* argv[])
```

Comment:
I get exhausted reading functions that are longer than 15 lines of code.  I am not able to parse what main is doing.  You should refactor and main should be nothing by a try/catch into some static function that calls other static functions.  The code itself should be understandable from the just reading what functions are called provided you use descriptive function names.

Reply:
I refactored as suggested. please suggest if any further changes would be required.